### PR TITLE
[CIS-1423] Custom user events are now properly decoded

### PR DESCRIPTION
**Pull request by bielikb**
### 🔗 Issue Link
CIS-1423

### 🎯 Goal

Parse custom user events.

### 🎨 Changes

* `UnknownEvent` is now deprecated, use `UnknownChannelEvent` instead.
* Adds new `UnknownUserEvent` that models custom user event.
* Extensive test suite.

### ☑️ Checklist

- [x] I have signed the [Stream CLA](https://docs.google.com/forms/d/e/1FAIpQLScFKsKkAJI7mhCr7K9rEIOpqIDThrWxuvxnwUq2XkHyG154vQ/viewform) (required)
- [x] This change follows zero ⚠️ policy (required)
- [x] Changelog is updated with client-facing changes
- [x] New code is covered by unit tests
- [ ] Affected documentation updated (docusaurus, tutorial, CMS (task created)

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -14,19 +14,22 @@ _December 15, 2021_
...
 - There is a new `ChatMessageListDateSeparatorView` component that should be used instead of the `ChatMessageListScrollOverlayView` if the goal is customize the styling of the date separator. Read [here](https://getstream.io/chat/docs/sdk/ios/uikit/components/message/#date-separators) for more details.
+- `UnknownEvent` is now deprecated, use `UnknownChannelEvent` or `UnknownUserEvent` instead. [#1695](https://github.com/GetStream/stream-chat-swift/pull/1695).
 
 ### 🐞 Fixed
-- Fix `stopTyping` can be called on `TypingEventSender` after calling `startTyping` [#1649](https://github.com/GetStream/stream-chat-swift/issues/1649)
-- Reactions no longer cover the text in message bubble [#1666](https://github.com/GetStream/stream-chat-swift/pull/1666)
-- Fix `error` type messages rendered as user's messages and interactive [#1672](https://github.com/GetStream/stream-chat-swift/issues/1672)
-- Fix `ChannelListController` makes one redundant API call [#1687](https://github.com/GetStream/stream-chat-swift/issues/1687)
-- Safely access indexes of collections [#1692](https://github.com/GetStream/stream-chat-swift/pull/1692)
+
```
Comment (evsaev) [inline CHANGELOG.md:20]:
```suggestion
```

File: Sources/StreamChat/APIClient/Endpoints/Payloads/UnknownChannelEvent.swift [hunk 2]
```diff
@@ -20,12 +20,12 @@ public struct UnknownEvent: Event, Hashable {
     public let createdAt: Date
     
     /// A dictionary with custom fields.
```
Comment (evsaev) [inline Sources/StreamChat/APIClient/Endpoints/Payloads/UnknownChannelEvent.swift:22]:
```suggestion
    /// A dictionary containing the entire event JSON.
```

File: Sources/StreamChat/APIClient/Endpoints/Payloads/UnknownUserEvent.swift [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+import Foundation
+
+/// An event type SDK fallbacks to if incoming event was failed to be
+/// decoded as a system event.
+public struct UnknownUserEvent: Event, Hashable {
+    /// An event type.
+    public let type: EventType
+    
+    /// A user identifier the event is sent by.
```
Comment (evsaev) [inline Sources/StreamChat/APIClient/Endpoints/Payloads/UnknownUserEvent.swift:16]:
```suggestion
    /// A user the event is triggered for.
```

File: Sources/StreamChat/APIClient/Endpoints/Payloads/UnknownUserEvent.swift [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    public let userId: UserId
+    
+    /// An event creation date.
+    public let createdAt: Date
+    
+    /// A dictionary with custom fields.
```
Comment (evsaev) [inline Sources/StreamChat/APIClient/Endpoints/Payloads/UnknownUserEvent.swift:22]:
```suggestion
    /// A dictionary containing the entire event JSON.
```

**Review by evsaev — APPROVED** [review]
LGTM 👍 Just some nits mostly about the docs

**Comment by bielikb** [comment]
> LGTM 👍 Just some nits mostly about the docs

Thanks for the docs suggestions. I applied them ;)

**Review by evsaev — APPROVED** [review]

File: CHANGELOG.md [hunk 5]
```diff
@@ -9,18 +9,22 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
...
 - There is a new `ChatMessageListDateSeparatorView` component that should be used instead of the `ChatMessageListScrollOverlayView` if the goal is customize the styling of the date separator. Read [here](https://getstream.io/chat/docs/sdk/ios/uikit/components/message/#date-separators) for more details.
+- `UnknownEvent` is now deprecated, use `UnknownChannelEvent` or `UnknownUserEvent` instead. [#1695](https://github.com/GetStream/stream-chat-swift/pull/1695).
 
 ### 🐞 Fixed
-- Fix `stopTyping` can be called on `TypingEventSender` after calling `startTyping` [#1649](https://github.com/GetStream/stream-chat-swift/issues/1649)
-- Reactions no longer cover the text in message bubble [#1666](https://github.com/GetStream/stream-chat-swift/pull/1666)
-- Fix `error` type messages rendered as user's messages and interactive [#1672](https://github.com/GetStream/stream-chat-swift/issues/1672)
-- Fix `ChannelListController` makes one redundant API call [#1687](https://github.com/GetStream/stream-chat-swift/issues/1687)
+
```
Comment (bielikb) [inline CHANGELOG.md:15]:
```suggestion
```

**Review by evsaev — DISMISSED** [review]

**Review by tbarbugli — CHANGES_REQUESTED** [review]
- Can we add a test for both events that decode a snapshot (json file)?
- Do we support custom data parsing for both (either way good to add a json file for that and ensure we cover that with a test)

**Comment by bielikb** [comment]
> * Can we add a test for both events that decode a snapshot (json file)?
> * Do we support custom data parsing for both (either way good to add a json file for that and ensure we cover that with a test)

Both cases are covered with tests. snapshots are inline == I followed the same approach that was already in place.

**Comment by bielikb** [comment]
@tbarbugli can you review the PR? Thanks.

**Pull request merged**